File: tests/c/test_version.c
#include <stdio.h>
#include <string.h>
#include <oti/oti.h>

int main(void) {
    int major = -1, minor = -1, patch = -1;
    char expected[64];

    /* The runtime accessor must agree with the compile-time macro. */
    if (strcmp(oti_version(), OTI_VERSION_STRING) != 0) {
        fprintf(stderr, "Version test failed: library reports %s, header declares %s\n",
                oti_version(), OTI_VERSION_STRING);
        return 1;
    }

    /* The numeric components must reconstruct the version string. */
    oti_version_numbers(&major, &minor, &patch);
    snprintf(expected, sizeof(expected), "%d.%d.%d", major, minor, patch);

    if (strcmp(oti_version(), expected) != 0) {
        fprintf(stderr, "Version test failed: components give %s, string is %s\n",
                expected, oti_version());
        return 1;
    }

    /* The packed version must match the encoding of its own components. */
    if (oti_version_number() != OTI_VERSION_ENCODE(major, minor, patch)) {
        fprintf(stderr, "Version test failed: packed version %d does not encode %s\n",
                oti_version_number(), expected);
        return 1;
    }

    /* NULL outputs are documented as accepted. */
    oti_version_numbers(NULL, NULL, NULL);

    printf("C version tests passed successfully (OTIlib %s).\n", oti_version());
    return 0;
}

File: tests/fortran/test_version.f90
PROGRAM test_version

    USE oti_version

    IMPLICIT NONE

    CHARACTER(LEN=32) :: rebuilt
    INTEGER           :: packed

    ! The version components must reconstruct the version string.
    WRITE(rebuilt, '(I0,".",I0,".",I0)') OTI_VERSION_MAJOR, OTI_VERSION_MINOR, OTI_VERSION_PATCH

    IF (TRIM(rebuilt) /= OTI_VERSION_STRING) THEN
        WRITE(*,*) "Version test failed: components give ", TRIM(rebuilt), &
                   " but string is ", OTI_VERSION_STRING
        STOP 1
    END IF

    ! The packed version must match the encoding used by the C and Python bindings.
    packed = OTI_VERSION_MAJOR * 10000 + OTI_VERSION_MINOR * 100 + OTI_VERSION_PATCH

    IF (OTI_VERSION_NUMBER /= packed) THEN
        WRITE(*,*) "Version test failed: packed version does not encode the components"
        STOP 1
    END IF

    WRITE(*,*) "Fortran version test passed successfully (OTIlib ", OTI_VERSION_STRING, ")."

END PROGRAM test_version
